File: backend-go/internal/middleware/validation.go
package middleware

import (
	"regexp"

	"github.com/gin-gonic/gin"
	"github.com/phitonias/streak/internal/utils"
)

var (
	usernameRegex = regexp.MustCompile(`^[a-zA-Z0-9_]{3,50}$`)
	emailRegex    = regexp.MustCompile(`^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$`)
	passwordRegex = regexp.MustCompile(`^(?=.*[a-z])(?=.*[A-Z])(?=.*\d).{8,}$`)
)

type RegisterRequest struct {
	Username        string   `json:"username" binding:"required"`
	Email           string   `json:"email" binding:"required"`
	Password        string   `json:"password" binding:"required"`
	Role            string   `json:"role" binding:"required"`
	DisplayName     string   `json:"displayName"`
	SportCategories []string `json:"sportCategories"`
}

type LoginRequest struct {
	Username string `json:"username" binding:"required"`
	Password string `json:"password" binding:"required"`
}

type CreateStreamRequest struct {
	Title         string   `json:"title" binding:"required,min=1,max=200"`
	Description   string   `json:"description"`
	SportCategory string   `json:"sportCategory" binding:"required"`
	Tags          []string `json:"tags"`
}

type UpdateStreamRequest struct {
	Title         string   `json:"title"`
	Description   string   `json:"description"`
	SportCategory string   `json:"sportCategory"`
	Tags          []string `json:"tags"`
	ThumbnailURL  string   `json:"thumbnailUrl"`
}

type UpdateProfileRequest struct {
	DisplayName     string   `json:"displayName"`
	Bio             string   `json:"bio"`
	AvatarURL       string   `json:"avatarUrl"`
	BannerURL       string   `json:"bannerUrl"`
	SportCategories []string `json:"sportCategories"`
}

func ValidateRegister(c *gin.Context) {
	var req RegisterRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.ErrorResponse(c, 400, "Invalid request body")
		c.Abort()
		return
	}

	// Validate username
	if !usernameRegex.MatchString(req.Username) {
		utils.ErrorResponse(c, 400, "Username must be 3-50 characters and contain only letters, numbers, and underscores")
		c.Abort()
		return
	}

	// Validate email
	if !emailRegex.MatchString(req.Email) {
		utils.ErrorResponse(c, 400, "Invalid email address")
		c.Abort()
		return
	}

	// Validate password
	if !passwordRegex.MatchString(req.Password) {
		utils.ErrorResponse(c, 400, "Password must be at least 8 characters with uppercase, lowercase, and number")
		c.Abort()
		return
	}

	// Validate role
	if req.Role != "athlete" && req.Role != "follower" {
		utils.ErrorResponse(c, 400, "Role must be athlete or follower")
		c.Abort()
		return
	}

	c.Set("validatedRequest", req)
	c.Next()
}

File: backend/src/middleware/validation.ts
import { Request, Response, NextFunction } from 'express';
import { validationResult, body, param, query } from 'express-validator';
import { SPORT_CATEGORIES } from '../types';

export const validate = (req: Request, res: Response, next: NextFunction): void => {
  const errors = validationResult(req);
  if (!errors.isEmpty()) {
    res.status(400).json({
      success: false,
      error: 'Validation failed',
      details: errors.array(),
    });
    return;
  }
  next();
};

// Auth validation rules
export const registerValidation = [
  body('username')
    .trim()
    .isLength({ min: 3, max: 50 })
    .matches(/^[a-zA-Z0-9_]+$/)
    .withMessage('Username must be 3-50 characters and contain only letters, numbers, and underscores'),
  body('email').trim().isEmail().normalizeEmail().withMessage('Invalid email address'),
  body('password')
    .isLength({ min: 8 })
    .matches(/^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)/)
    .withMessage('Password must be at least 8 characters with uppercase, lowercase, and number'),
  body('role').isIn(['athlete', 'follower']).withMessage('Role must be athlete or follower'),
  body('displayName').optional().trim().isLength({ max: 100 }),
  validate,
];

export const loginValidation = [
  body('username').trim().notEmpty().withMessage('Username is required'),
  body('password').notEmpty().withMessage('Password is required'),
  validate,
];

// Stream validation rules
export const createStreamValidation = [
  body('title').trim().isLength({ min: 1, max: 200 }).withMessage('Title must be 1-200 characters'),
  body('description').optional().trim().isLength({ max: 1000 }),
  body('sportCategory').isIn(SPORT_CATEGORIES).withMessage('Invalid sport category'),
  body('tags').optional().isArray().withMessage('Tags must be an array'),
  body('tags.*').optional().trim().isLength({ max: 50 }),
  validate,
];

export const updateStreamValidation = [
  body('title').optional().trim().isLength({ min: 1, max: 200 }),
  body('description').optional().trim().isLength({ max: 1000 }),
  body('sportCategory').optional().isIn(SPORT_CATEGORIES),
  body('tags').optional().isArray(),
  body('tags.*').optional().trim().isLength({ max: 50 }),
  validate,
];

// User profile validation
export const updateProfileValidation = [
  body('displayName').optional().trim().isLength({ max: 100 }),
  body('bio').optional().trim().isLength({ max: 500 }),
  body('sportCategories').optional().isArray(),
  body('sportCategories.*').optional().isIn(SPORT_CATEGORIES),
  validate,
];

// Pagination validation
export const paginationValidation = [
  query('page').optional().isInt({ min: 1 }).toInt(),
  query('limit').optional().isInt({ min: 1, max: 100 }).toInt(),
  validate,
];

// UUID validation
export const uuidValidation = [
  param('id').isUUID().withMessage('Invalid UUID format'),
  validate,
];
